packages/optimistic-oracle/src/keeper.js

```diff
@@ -0,0 +1,250 @@
+  /**
+   * @notice Constructs new OO Keeper bot.
+   * @param {Object} logger Module used to send logs.
+   * @param {Object} optimisticOracleClient Module used to query OO information on-chain.
+   * @param {Object} gasEstimator Module used to estimate optimal gas price with which to send txns.
+   * @param {String} account Ethereum account from which to send txns.
+   * @param {Object} defaultPriceFeedConfig Default configuration to construct all price feed objects.
+   * @param {Object} [ooKeeperConfig] Contains fields with which constructor will attempt to override defaults.
+   */
+  constructor({ logger, optimisticOracleClient, gasEstimator, account, defaultPriceFeedConfig, ooKeeperConfig }) {
+    this.logger = logger;
+    this.account = account;
+    this.optimisticOracleClient = optimisticOracleClient;
+    this.web3 = this.optimisticOracleClient.web3;
+    this.defaultPriceFeedConfig = defaultPriceFeedConfig;
+
+    // Gas Estimator to calculate the current Fast gas rate.
+    this.gasEstimator = gasEstimator;
+
+    // Multiplier applied to Truffle's estimated gas limit for a transaction to send.
+    this.GAS_LIMIT_BUFFER = 1.25;
+
+    this.ooContract = this.optimisticOracleClient.oracle;
+
+    // Cached mapping of identifiers to pricefeed classes
+    this.priceFeedCache = {};
+
+    // Helper functions from web3.
+    this.BN = this.web3.utils.BN;
+    this.toBN = this.web3.utils.toBN;
+    this.toWei = this.web3.utils.toWei;
+    this.fromWei = this.web3.utils.fromWei;
+    this.utf8ToHex = this.web3.utils.utf8ToHex;
+    this.hexToUtf8 = this.web3.utils.hexToUtf8;
+
+    // Default config settings. Liquidator deployer can override these settings by passing in new
+    // values via the `ooKeeperConfig` input object. The `isValid` property is a function that should be called
+    // before resetting any config settings. `isValid` must return a Boolean.
+    const defaultConfig = {
+      txnGasLimit: {
+        // `txnGasLimit`: Gas limit to set for sending on-chain transactions.
+        value: 9000000, // Can see recent averages here: https://etherscan.io/chart/gaslimit
+        isValid: x => {
+          return x >= 6000000 && x < 15000000;
+        }
+      }
+    };
+
+    // Validate and set config settings to class state.
+    const configWithDefaults = createObjectFromDefaultProps(ooKeeperConfig, defaultConfig);
+    Object.assign(this, configWithDefaults);
+  }
+
+  async update() {
+    await Promise.all([this.optimisticOracleClient.update(), this.gasEstimator.update()]);
+
+    // Increase allowances for all relevant collateral currencies.
+    // TODO: Consider whether this should happen in a separate function (i.e. `setAllowances`) or within
+    // the `sendProposals()` method.
+    await this._setAllowances();
+  }
+
+  // Sets allowances for all collateral currencies used in unproposed price requests
+  async _setAllowances() {
+    // TODO: Should we do this also for currencies from undisputedProposals?
```

Comment:
I think it's fine to do bulk approval like this for now. When we move over to using DSProxy we can do the approval in the same tx that we sent the request/proposal.

Reply:
Yeah good call. Bulk approval is an upfront cost that I think we should pay so we don't have any approval errors come proposal/dispute time